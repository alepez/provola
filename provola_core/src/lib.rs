mod test_case;
mod testable;
mod report;
mod code;
mod error;
mod collection;
mod test_suite;
mod named;
mod report_future;

pub use test_case::TestCase;
pub use testable::Testable;
pub use report::Report;
pub use report::FailureDetails;
pub use named::Named;
pub use test_suite::TestSuite;