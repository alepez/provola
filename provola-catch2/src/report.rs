use provola_core::Failure as CoreFailure;
use provola_core::Report as CoreReport;
use provola_core::TestCase as CoreTestCase;
use provola_core::TestSuite as CoreTestSuite;
use serde::{Deserialize, Serialize};

type Duration = String;

pub type Name = String;
pub type Status = String;
pub type ClassName = String;
pub type FailureType = String;
pub type Message = String;

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct Report {
    #[serde(rename = "testsuite")]
    pub testsuites: Vec<TestSuite>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct TestSuite {
    // // TODO system-err
    // // TODO system-out
    #[serde(rename = "testcase", default)]
    pub testcases: Vec<TestCase>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct TestCase {
    pub classname: ClassName,
    pub name: Name,
    pub status: Status,
    pub time: Duration,
    #[serde(rename = "failure", default)]
    pub failures: Vec<Failure>,
}

#[derive(Serialize, Deserialize, Debug, Default)]
pub struct Failure {
    #[serde(rename = "type")]
    pub ttype: FailureType,
    pub message: Message,
}

impl From<Report> for CoreReport {
    fn from(x: Report) -> Self {
        // TODO
        CoreReport {
            ..Default::default()
        }
    }
}

#[cfg(test)]
mod tests {
    use std::{fs::File, io::BufReader};

    use super::*;

    #[test]
    fn parse_xml_report() {
        let path = "examples/data/test_report.xml";
        let file = File::open(path).unwrap();
        let reader = BufReader::new(file);
        let u: Report = serde_xml_rs::from_reader(reader).unwrap();
        insta::assert_debug_snapshot!(&u);
    }
}
