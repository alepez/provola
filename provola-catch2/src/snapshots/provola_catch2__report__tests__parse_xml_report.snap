---
source: provola-catch2/src/report.rs
expression: "&u"

---
Report {
    testsuites: [
        TestSuite {
            testcases: [
                TestCase {
                    classname: "example.global",
                    name: "Foo",
                    status: "run",
                    time: "0.000",
                    failures: [
                        Failure {
                            ttype: "REQUIRE",
                            message: "false",
                        },
                    ],
                },
                TestCase {
                    classname: "example.Bar",
                    name: "Bar1",
                    status: "run",
                    time: "0.000",
                    failures: [],
                },
                TestCase {
                    classname: "example.Bar",
                    name: "Bar2",
                    status: "run",
                    time: "0.000",
                    failures: [],
                },
            ],
        },
    ],
}
